src/library/scala/collection/Iterable.scala

```diff
@@ -228,6 +228,88 @@ trait IterableOps[+A, +CC[_], +C] extends Any with IterableOnce[A] with Iterable
   /** A view over the elements of this collection. */
   def view: View[A] = View.fromIteratorProvider(() => iterator)
 
+  /** Compares the size of this $coll to a test value.
+    *
+    *   @param   _size the test value that gets compared with the size.
+    *   @return  A value `x` where
+    *   {{{
+    *        x <  0       if this.size <  _size
+    *        x == 0       if this.size == _size
+    *        x >  0       if this.size >  _size
+    *   }}}
+    *  The method as implemented here does not call `size` directly; its running time
+    *  is `O(size min _size)` instead of `O(size)`. The method should be overwritten
+    *  if computing `size` is cheap.
+    */
+  def sizeCompare(_size: Int): Int = {
+    if (_size < 0) 1
+    else {
+      val known = knownSize
+      if (known >= 0) Integer.compare(known, _size)
+      else {
+        var i = 0
+        val it = iterator
+        while (it.hasNext) {
+          if (i == _size) return if (it.hasNext) 1 else 0
+          it.next()
+          i += 1
+        }
+        i - _size
+      }
+    }
+  }
+
+  /** Returns a value class containing operations for comparing the size of this $coll to a test value.
+    *
+    * These operations are implemented in terms of [[sizeCompare(Int) `sizeCompare(Int)`]], and
+    * allow the following more readable usages:
+    *
+    * {{{
+    * this.sizeIs < size     // this.sizeCompare(size) < 0
+    * this.sizeIs <= size    // this.sizeCompare(size) <= 0
+    * this.sizeIs == size    // this.sizeCompare(size) == 0
+    * this.sizeIs != size    // this.sizeCompare(size) != 0
+    * this.sizeIs >= size    // this.sizeCompare(size) >= 0
+    * this.sizeIs > size     // this.sizeCompare(size) > 0
+    * }}}
+    */
+  @inline final def sizeIs: IterableOps.SizeCompareOps = new IterableOps.SizeCompareOps(this)
+
+  /** Compares the size of this $coll to the size of another `Iterable`.
+    *
+    *   @param   that the `Iterable` whose size is compared with this $coll's size.
+    *   {{{
+    *        x <  0       if this.size <  that.size
+    *        x == 0       if this.size == that.size
+    *        x >  0       if this.size >  that.size
+    *   }}}
+    *  The method as implemented here does not call `size` directly; its running time
+    *  is `O(this.size min that.size)` instead of `O(this.size + that.size)`.
+    *  The method should be overwritten if computing `size` is cheap.
+    */
+  def compareSizesWith(that: Iterable[_]): Int = {
```

Comment:
What about using the singular for “size” here: `compareSizeWith`?

Reply:
IANAG(rammarian), but I thought that "sizes" makes more sense, since I am comparing the size of this to the size of that (i.e. two sizes), rather than the size of this to that (which would be one size, and also nonsense). I am happy to be be corrected by someone that plural is not more grammatically correct in this case.

Reply:
Do you feel more comfortable with `compareBySizeWith`?

Reply:
Neither of the `With` ones are great because we normally use `With` to mean we're using a different comparator (e.g. `sortWith`).  `compareSizeTo` is fine (shorthand for "compare my size to `that`'s size").  Or, as we did with `compare`, we could simply drop the `To`--not `compareTo`, just `compare`, so not `compareSizeTo` but just `compareSize`.

The plural works too (shorthand for "compare our respective sizes"), but since the non-plural version has a sensible meaning I'd go for it instead.  Having to remember random s's is annoying.

Reply:
honestly, that makes me more inclined to call it `sizeCompare`. I think it makes more sense grammatically than `compareSize` (which I still opine ought to use the plural), and is also consistent with the other `sizeCompare` method. I think that `compareSizesWith` is sufficiently different from `sizeCompare` that one would be unlikely to mix them up, but having both `sizeCompare` and `compareSize` would quite confusing.

Reply:
Overloading `sizeCompare` looks good to me.